### ISSUE TITLE
Investigate GuessZoneIdByTransitionsUncached performance

### ISSUE THREAD

**Issue by jskeet** [nodatime/nodatime#686]
This one parameterized test takes ~30 seconds at the moment, which seems way too long for running only ~132 test cases.

To investigate:
- See whether this is a regression from 1.3
- See whether most of that time is in DateTimeZone

**Comment by jskeet** [comment]
Observations:

- This is about twice as fast in release as in debug (at least in net451)
- It's not just guessing the transitions - it's then checking for correctness for every day over a 2 year period for the resulting time zone. But commenting out that part of the test doesn't change anything
- We're doing a lot of work each time the method is called, that isn't actually specific to the time zone being checked.
- There may be a regression due to needing to call `Instant.ToDateTimeUtc` once per candidate, per instant being checked - which is a 593x316 matrix
- If we need a hit score of at least 70%, we could reject zones as soon as they hit a particular failure rate

Basically, there's a lot we could optimize here. Will look at how reasonable it is to do so.

**Comment by jskeet** [comment]
Cost of adding in five extra calls within the inner `Count` loop, in release mode (which normally takes about 11 seconds):

- `candidate.GetUtcOffset(instant)` - about 0.8s
- `instant.ToDateTimeUtc()` - about 0.9s
- `zone.GetUtcOffset(...)` - about 15s (!!!)

Further observation - we're making the expensive `TimeZoneInfo` repeatedly for the same instants, over and over again. If we remove that redundancy, that should be the bulk of the call. Additionally, that *does* change with each call, so there's not a lot of point in caching things, probably.

**Comment by jskeet** [comment]
Okay, not quite as big a benefit as expected: reduced from 11s to 7s. Great, but we can do better.
Removing the `candidate.GetUtcOffset` call reduces the time by another ~2s, which makes little sense to me given the above comment. Hmm.

**Comment by jskeet** [comment]
(Calling `DateTimeZone.GetUtcOffset` multiple times will hit the cache of course... but we should be hitting the cache anyway due to the test being across all BCL zones. Maybe it's a processor cache difference?)

Now adding code to give up as soon as we know we should...

**Comment by jskeet** [comment]
Ha - fresh observation... this line:

```csharp
var candidates = CanonicalIdMap.Values.Select(ForId).ToList()
```

... takes about half the time of the test! That should be somewhat expected, given that a time zone *source* doesn't implement caching.

Next observation: this method is only called in non-test code via `GetSystemDefault`, which goes via a cache. So unless the user changes local time zone frequently, we're only going to get to this code once... so the only benefit from any further optimization would be to speed up the tests. I don't think we need to go there.

Final result: let's avoid being completely gratuitous with `DateTimeZone`, and early-out when we've seen more failures than we want, but avoid adding even more caching.

**Comment by jskeet** [comment]
Just worked out how to improve this massively. Much of the problem is that we reload the time zones from raw data on each iteration. If we move this line of code:

```csharp
var candidates = CanonicalIdMap.Values.Select(ForId).ToList();
```

to `GuessZoneIdByTransitions` (which provides caching), and pass it into `GuessZoneIdByTransitionsUncached`, then the tests can pass in the same list for each BCL zone.

Not only does that save us the time of loading, but it means that after the first test, the caches will all be populated.
